### PR TITLE
Waterfall: log commands before executing them

### PR DESCRIPTION
**Pull request by dschuff**

### CODE REVIEW

**Comment by jfbastien** [comment]
Could you instead create a `subprocess` wrapper (say `proc.py`) and wrap the functions we use from subprocess? Seems better than duplicating the same function ;-)

**Comment by dschuff** [comment]
eh, there was actually no duplication since the 2 files used different subsets of subprocess, but I came up with something more clever (by which I mean probably worse?) instead.

File: buildbot/build.py [hunk 1]
```diff
@@ -22,6 +22,8 @@
 import multiprocessing
 import urllib2
 
+# Use proc instead of subprocess to get logged subprocess invocation
+import proc
```
Comment (jfbastien) [inline buildbot/build.py]: Sort imports, drop the comment (it should be in proc.py instead).

File: buildbot/build.py [hunk 2]
```diff
@@ -31,7 +33,7 @@
 WORK_DIR = os.path.join(SCRIPT_DIR, 'work')
 
 CLOUD_STORAGE_BASE_URL = 'https://storage.googleapis.com/'
-CLOUD_STORAGE_PATH = 'wasm-llvm/builds/'
+CLOUD_STORAGE_PATH = 'Wasm-llvm/builds/'
```
Comment (jfbastien) [inline buildbot/build.py]: ?

File: buildbot/build.py [hunk 3]
```diff
@@ -210,13 +211,13 @@ def Archive(name, tar):
 
 def GitRemoteUrl(cwd, remote):
   """Get the URL of a remote."""
-  return subprocess.check_output(['git', 'config', '--get', 'remote.%s.url' %
+  return proc.check_output(['git', 'config', '--get', 'remote.%s.url' %
                                   remote], cwd=cwd).strip()
```
Comment (jfbastien) [inline buildbot/build.py]: Indentation, here and elsewhere.

File: buildbot/build.py (hunk 1 above)
Comment (dschuff) [inline buildbot/build.py]: The comment is in the importers of proc to keep someone from adding an import of subprocess.

File: buildbot/build.py (hunk 2 above)
Comment (dschuff) [inline buildbot/build.py]: thanks, emacs.

File: buildbot/proc.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+# This module is intended to be a drop-in replacement for the standard
+# subprocess module, with the difference that it logs commands before it runs
+# them. Everything not overriden should pass through to the subprocess module
+# via the import trick below.
+
+# Imports subprocess in its own namespace so we can always refer directly to
+# its attributes
```
Comment (jfbastien) [inline buildbot/proc.py]: Period at end of sentence

File: buildbot/proc.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import subprocess
+# Imports all of subprocess into the current namespace, effectively re-exporting
+# everything.
+from subprocess import *
+
+
+# Now we can override any parts of subprocess we want, while leaving the rest
```
Comment (jfbastien) [inline buildbot/proc.py]: Ditto

File: buildbot/build.py (hunk 1 above)
Comment (dschuff) [inline buildbot/build.py]: Also I assume the imports are the way they are to distinguish between system imports and user imports? I'll check PEP8 on that.

File: buildbot/proc.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+def check_call(cmd, **kwargs):
+  print 'Running command:', cmd
```
Comment (jfbastien) [inline buildbot/proc.py]: Something like this (untested):
```
' '.join('"' + c + '"' if ' ' in c else c for c in cmd)
```
So you can just copy / paste the command?

File: buildbot/testing.py [hunk 7]
```diff
@@ -15,11 +15,12 @@
...
 
 import difflib
-import os
-import os.path
 import math
 import multiprocessing
-import subprocess
+import os
+import os.path
+# Use proc instead of subprocess to get logged subprocess invocation
```
Comment (jfbastien) [inline buildbot/testing.py]: Drop comment here too.

File: buildbot/build.py (hunk 1 above)
Comment (dschuff) [inline buildbot/build.py]: Sorted the local imports.

**Comment by jfbastien** [comment]
Clever, I like it. A few nits, rest is good.

File: buildbot/build.py (hunk 3 above)
Comment (dschuff) [inline buildbot/build.py]: Done.

File: buildbot/proc.py (hunk 5 above)
Comment (dschuff) [inline buildbot/proc.py]: Done.

File: buildbot/proc.py (hunk 6 above)
Comment (dschuff) [inline buildbot/proc.py]: We went through this exact same debate with NaCl's buildbot scripts, and those who wanted to print the python list (just to be extra-explicit about what was happening) won. I'm happy either way though. Maybe do both?

Comment (dschuff) [inline buildbot/proc.py]: Nevermind, just went with the copyable version.

**Pull request merged**